contrib/babelfishpg_tsql/src/hooks.c

```diff
@@ -866,6 +905,39 @@ pltsql_ExecutorStart(QueryDesc *queryDesc, int eflags)
 			queryDesc->instrument_options |= INSTRUMENT_WAL;
 	}
 
+	/* In TSQL dialect the RTE permissions might need to be checked against session user. */
+	if (sql_dialect == SQL_DIALECT_TSQL && queryDesc->plannedstmt != NULL)
+	{
+		ListCell   *lc;
+
+		foreach(lc, queryDesc->plannedstmt->permInfos)
+		{
+			RTEPermissionInfo	*perminfo = lfirst_node(RTEPermissionInfo, lc);
+			Oid             	relOid = perminfo->relid;
+
+			if (OidIsValid(relOid))
+			{
+				Oid schema_id = get_rel_namespace(relOid);
+
+				if (OidIsValid(schema_id))
+				{
+					char *nspname = get_namespace_name(schema_id);
+
+					/*
+					 * Check if relation's schema is from a different logical database and
+					 * it is not a shared schema. If yes, then update checkAsUser to session
+					 * user to allow cross database access.
+					 */
+					if (nspname != NULL && !is_shared_schema(nspname) &&
+						!is_schema_from_db(schema_id, get_cur_db_id()))
+						perminfo->checkAsUser = GetSessionUserId();
+					if (nspname)
+						pfree(nspname);
+				}
+			}
+		}
+	}
```

Comment:
After this changes, Can we remove any redundant logic from `pl_exec.c` or `pl_exec-2.c`?

Reply:
Yep, removed it.